library/core/src/iter/adapters/dedup.rs

```diff
@@ -0,0 +1,168 @@
+/// See its documentation for more.
+///
+/// [`Iterator::dedup_by_key`]: Iterator::dedup_by_key
+#[unstable(feature = "iter_dedup", reason = "recently added", issue = "83748")]
+#[derive(Debug, Clone, Copy)]
+pub struct ByKey<F> {
+    key: F,
+}
+
+impl<F> ByKey<F> {
+    pub(crate) fn new(key: F) -> Self {
+        Self { key }
+    }
+}
+
+#[unstable(feature = "iter_dedup", reason = "recently added", issue = "83748")]
+impl<F, T, K> FnOnce<(&T, &T)> for ByKey<F>
+where
+    F: FnMut(&T) -> K,
+    K: PartialEq,
+{
+    type Output = bool;
+
+    extern "rust-call" fn call_once(mut self, args: (&T, &T)) -> Self::Output {
+        (self.key)(args.0) == (self.key)(args.1)
+    }
+}
+
+#[unstable(feature = "iter_dedup", reason = "recently added", issue = "83748")]
+impl<F, T, K> FnMut<(&T, &T)> for ByKey<F>
+where
+    F: FnMut(&T) -> K,
+    K: PartialEq,
+{
+    extern "rust-call" fn call_mut(&mut self, args: (&T, &T)) -> Self::Output {
+        (self.key)(args.0) == (self.key)(args.1)
+    }
+}
+
+/// A zero-sized type for checking partial equality.
+///
+/// This struct acts exactly like the function [`PartialEq::eq`], but its
+/// type is always known during compile time.
+///
+/// This `struct` is created by [`Iterator::dedup`].
+/// See its documentation for more.
+///
+/// [`Iterator::dedup`]: Iterator::dedup
+#[unstable(feature = "iter_dedup", reason = "recently added", issue = "83748")]
+#[derive(Debug, Clone, Copy)]
+pub struct ByPartialEq;
+
+impl ByPartialEq {
+    pub(crate) fn new() -> Self {
+        Self
+    }
+}
+
+#[unstable(feature = "iter_dedup", reason = "recently added", issue = "83748")]
+impl<T: PartialEq> FnOnce<(&T, &T)> for ByPartialEq {
+    type Output = bool;
+
+    extern "rust-call" fn call_once(self, args: (&T, &T)) -> Self::Output {
+        args.0 == args.1
+    }
+}
+
+#[unstable(feature = "iter_dedup", reason = "recently added", issue = "83748")]
+impl<T: PartialEq> FnMut<(&T, &T)> for ByPartialEq {
+    extern "rust-call" fn call_mut(&mut self, args: (&T, &T)) -> Self::Output {
+        args.0 == args.1
+    }
+}
+
+/// An iterator that removes all but the first of consecutive elements in a
+/// given iterator satisfying a given equality relation.
+///
+/// This `struct` is created by [`Iterator::dedup`], [`Iterator::dedup_by`]
+/// and [`Iterator::dedup_by_key`]. See its documentation for more.
+///
+/// [`Iterator::dedup`]: Iterator::dedup
+/// [`Iterator::dedup_by`]: Iterator::dedup_by
+/// [`Iterator::dedup_by_key`]: Iterator::dedup_by_key
+#[unstable(feature = "iter_dedup", reason = "recently added", issue = "83748")]
+#[derive(Debug, Clone, Copy)]
+pub struct Dedup<I, F>
+where
+    I: Iterator,
+{
+    inner: I,
+    same_bucket: F,
+    last: Option<I::Item>,
+}
+
+impl<I, F> Dedup<I, F>
+where
+    I: Iterator,
+{
+    pub(crate) fn new(inner: I, same_bucket: F) -> Self {
+        let mut inner = inner;
+        Self { last: inner.next(), inner, same_bucket }
```

Comment:
Eagerly fetching the first element here seems surprising. Perhaps this be avoided by using `Peekable`?

Reply:
You're right, this might be surprising if the inner iterator has side effects. I'm not sure `peakable` is a good way to fix this, but I can move the initial call to `next` into `Dedup::next` no problem.

Reply:
I imagine itertools does it this way to make `Dedup` smaller, but I don't think it's good.

I think this would work? (inner is Peekable)
```rust
fn next() {
    let next = self.inner.next()?;
    while self.inner.next_if_eq(&self.next).is_some() {}
    next
}
```

Reply:
The closest I can get to your suggested implementation is this
```rust
fn next(&mut self) -> Option<Self::Item> {
    // ensure we don't go past the first `None`
    self.inner.peek()?;

    // get item to be returned
    let curr = self.inner.next()?;

    // prepare inner iterator for following call to `Self::next`
    let same_bucket = &mut self.same_bucket;
    while self
        .inner
        .next_if(|next| (same_bucket)(next, &curr))
        .is_some()
    {}

    Some(curr)
}
```
which isn't isn't as nice as I would have hoped.

It also seems to fail to derive the `Debug` and `Clone` impls and the debug output being less helpful for debugging.

Reply:
For now, I pushed a version that uses `Option<Option<I::Item>>` to distinguish never having called `next` on `inner` and getting `None` from `inner`